examples/hetero/recommender_system.py

```diff
@@ -0,0 +1,434 @@
+        for i, graph in enumerate(graphs):
+            graph[edge_name][attr] = data_splits[i]
+
+BATCH_SIZE = 512
+
+train_dataloader = LinkNeighborLoader(
+    data=train_data, num_neighbors=[5, 5, 5], neg_sampling_ratio=1,
+    edge_label_index=(('user', 'rates', 'movie'),
+                      train_data[('user', 'rates', 'movie')].edge_index),
+    edge_label_time=train_data[('user', 'rates', 'movie')].time,
+    batch_size=BATCH_SIZE, shuffle=True, time_attr='time')
+
+val_dataloader = LinkNeighborLoader(
+    data=val_data,
+    num_neighbors=[-1, -1, -1],
+    neg_sampling_ratio=1,
+    edge_label_index=(('user', 'rates', 'movie'),
+                      val_data[('user', 'rates', 'movie')].edge_index),
+    batch_size=BATCH_SIZE,
+    shuffle=True,
+)
+
+test_dataloader = LinkNeighborLoader(
+    data=test_data,
+    num_neighbors=[-1, -1, -1],
+    neg_sampling_ratio=1,
+    edge_label_index=(('user', 'rates', 'movie'),
+                      test_data[('user', 'rates', 'movie')].edge_index),
+    batch_size=BATCH_SIZE,
+    shuffle=True,
+)
+
+# We have an unbalanced dataset with many labels for rating 3 and 4, and very
+# few for 0 and 1. Therefore we use a weighted MSE loss.
+if args.use_weighted_loss:
+    weight = torch.bincount(data['user', 'movie'].edge_label)
+    weight = weight.max() / weight
+else:
+    weight = None
+
+
+def weighted_mse_loss(pred, target, weight=None):
+    weight = 1. if weight is None else weight[target].to(pred.dtype)
+    return (weight * (pred - target.to(pred.dtype)).pow(2)).mean()
+
+
+class GNNEncoder(torch.nn.Module):
+    def __init__(self, hidden_channels, out_channels):
+        super().__init__()
+        self.conv1 = SAGEConv((-1, -1), hidden_channels)
+        self.conv2 = SAGEConv((-1, -1), hidden_channels)
+        self.conv3 = SAGEConv((-1, -1), out_channels)
+
+    def forward(self, x, edge_index):
+        x = self.conv1(x, edge_index).relu()
+        x = self.conv2(x, edge_index).relu()
+        x = self.conv3(x, edge_index)
+        return x
+
+
+class EdgeDecoder(torch.nn.Module):
+    def __init__(self, hidden_channels):
+        super().__init__()
+        self.lin1 = Linear(2 * hidden_channels, hidden_channels)
+        self.lin2 = Linear(hidden_channels, 1)
+        self.user_to_item = Linear(hidden_channels, hidden_channels)
+        self.item_to_user = Linear(hidden_channels, hidden_channels)
+
+    def forward(self, z_dict, edge_label_index):
+        row, col = edge_label_index
+        u_to_i = self.user_to_item(z_dict['user'][row])
+        i_to_u = self.item_to_user(z_dict['movie'][col])
+
+        z = torch.cat([u_to_i, i_to_u], dim=-1)
+
+        z = self.lin1(z).relu()
+        z = self.lin2(z)
+        return z.view(-1)
+        # return z
+
+
+class Model(torch.nn.Module):
+    def __init__(self, num_users, num_movies, hidden_channels):
+        super().__init__()
+        self.user_emb = Embedding(num_users, hidden_channels, device=device)
```

Comment:
If you are using an `Embedding` here, then there is no need to add a one-hot matrix for user features above.

Reply:
Agreed. I have removed the one hot encoding. Thanks